bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,401 @@
+use std::{
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
+use s2n_tls_sys::*;
+
+use crate::error::{Error, Fallible};
+
+#[derive(Copy, Clone)]
+pub enum FingerprintType {
+    JA3,
+}
+
+// this is the size of the MD5 hash digest that is used for the JA3 fingerprint
+const MD5_HASH_SIZE: u32 = 16;
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+/// ClientHello is an opaque wrapper struct around `s2n_client_hello`. Note that
+/// the size of this type is not known, and as such it can only be used through
+/// references and pointers.
+///
+/// This implementation is motivated by the different memory_management required
+/// for different s2n_client_hello pointers. `s2n_client_hello_parse_message`
+/// returns a `*mut s2n_client_hello` which owns it's own data. This neatly fits
+///  the "smart pointer" pattern and can be represented as a `Box<T>`.
+///
+/// `s2n_connection_get_client_hello` returns a `*mut s2n_client_hello` which
+/// references memory owned by the connection, and therefore must not outlive
+/// the connection struct. This is best represented as a reference tied to the
+/// lifetime of the `Connection` struct.
+
+pub struct ClientHello(s2n_client_hello);
+
+// safety justifications
+// 1 - casting *s2n_client_hello <-> *ClientHello: a struct with only one field
+// has the same layout as that field
+// https://rust-lang.github.io/unsafe-code-guidelines/layout/structs-and-tuples.html#single-field-structs
+// 2 - casting *const s2n_client_hello -> *mut s2n_client_hello: This is safe as
+// long as the data is not actually mutated. As authors of s2n-tls, we know that
+// the get_hash and get_fingerprint methods do not mutate the data, and use mut
+// pointers as a matter of convention because it makes working with s2n_stuffers
+// and s2n_blobs easier.
+
+impl Deref for ClientHello {
+    type Target = s2n_client_hello;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for ClientHello {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
```

Comment:
Implementing this will allow external callers to access the raw functions associated with the client hello. Is this something we want? I think I'd prefer to have safe functions for everything so they don't need this.

Reply:
I agree that they shouldn't need access to it. I had implemented this so that it would be easier for us to work with the wrapper type, and didn't intend it for public consumption.

For them to access the raw functions, they'd have to pull in `s2n-tls-sys`, which we explicitly describe as not intended for direct consumption. And if they are that determined to use the client hello with raw functions, there are unsafe ways of bludgeoning the `ClientHello` into an `s2n_client_hello`. 

I don't have particularly strong feeling either way. I think it's a little bit nicer for our own implementation to have access to Deref, but can just switch these to private helper methods if you think that'd be better.